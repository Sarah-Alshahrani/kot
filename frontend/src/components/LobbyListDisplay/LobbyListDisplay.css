.game_room_details {
    justify-content: center;
    display: flex;
    flex-direction: row;
    border-width: thin
}