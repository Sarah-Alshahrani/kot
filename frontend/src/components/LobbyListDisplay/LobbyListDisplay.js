import React from "react";
import { Redirect } from "react-router-dom";
import "./LobbyListDisplay.css";
import Table from "react-bootstrap/Table";
import GameInstance from "../../services/gameService";

class LobbyListDisplay extends React.Component {
  constructor(props, context) {
    super(props, context);

    this.state = {
      redirect: false,
      username: props.currentUser,
      gameRoom: props.currentRoom,
      data: props.data
    };

    GameInstance.addGameListResponseCallback(
      this.gameListResponseHandler.bind(this)
    );

    this.waitForSocketConnection(() => {
      const request = this.createRequestGameListCommand(
        this.state.username,
        this.state.gameRoom
      );
      GameInstance.sendMessage(request);
    });
  }

  waitForSocketConnection(callback) {
    const component = this;
    setTimeout(function() {
      // Check if websocket state is OPEN
      if (GameInstance.state() === 1) {
        console.log("Lobby connection is made");
        callback();
        return;
      } else {
        console.log("wait for lobby connection...");
        component.waitForSocketConnection(callback);
      }
    }, 100); // wait 100 milisecond for the connection...
  }

  createRequestGameListCommand(user, room) {
    return {
      command: "request_game_list",
      user: user,
      room: room
    };
  }

  gameListResponseHandler(message) {
    const content = message.content;

    if (content === undefined || content === "") return;

    if (content instanceof Object) {
      let game_list = [];
      try {
        for (const game_key in content) {
          let game = content[game_key];
          let player_detail = JSON.parse(game);

          let player_list = [];
          for (const person_key in player_detail) {
            let player = player_detail[person_key];
            player_list.push(player.username);
          }

          game_list.push({
            room_name: game_key,
            users: player_list.join()
          });
        }
      } catch (e) {}

      if (game_list.length) {
        this.setState({ data: game_list });
      }
    }
  }

  setRedirect = e => {
    this.setState({
      gameRoom: e.target.id,
      redirect: true
    });
  };

  renderRedirect = () => {
    const user = this.state.username;
    const room = this.state.gameRoom;
    if (this.state.redirect) {
      return (
        <Redirect
          to={{
            pathname: "/game",
            state: {
              username: user,
              gameRoom: room
            }
          }}
        />
      );
    }
  };

  renderState() {
    if (this.state.data instanceof Array) {
      if (this.state.data.length) {
        return (
          <div>
            <Table striped bordered hover>
              <thead>
                <tr>
                  <th>Players</th>
                  <th>Room Name</th>
                  <th></th>
                </tr>
              </thead>
              <tbody>
                {this.state.data.map((entry, index) => (
                  <tr key={index}>
                    <td>{entry.users}</td>
                    <td>{entry.room_name}</td>
                    <td>
                      <a className="b" href="/game">
                        {this.renderRedirect()}
                        <button
                          id={entry.room_name}
                          onClick={this.setRedirect} // (this.state.gameRoom)
                          className="btn btn-secondary"
                          type="button"
                        >
                          Join Room {entry.room_name}
                        </button>
                      </a>
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </div>
        );
      }
    }
    return this.renderDefault();
  }

  renderDefault() {
    return (
      <div>
        <Table striped bordered hover>
          <thead>
            <tr>
              <th>Players</th>
              <th>Room Name</th>
              <th></th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>{this.state.username}</td>
              <td>{this.state.gameRoom}</td>
              <td>
                <a className="b" href="/game">
                  {this.renderRedirect()}
                  <button
                    id={this.state.gameRoom}
                    onClick={this.setRedirect} // (this.state.gameRoom)
                    className="btn btn-secondary"
                    type="button"
                  >
                    Join Room {this.state.gameRoom}
                  </button>
                </a>
              </td>
            </tr>
          </tbody>
        </Table>
      </div>
    );
  }

  render() {
    return this.renderState();
  }
}
export default LobbyListDisplay;
