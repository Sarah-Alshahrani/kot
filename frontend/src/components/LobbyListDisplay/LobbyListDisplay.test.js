import React from "react";
import { shallow } from "enzyme";

import LobbyListDisplay from "./LobbyListDisplay";

describe("Verify LobbyListDisplay layout template", () => {
  const test_player_details = [
    {
      users: ["edward", "marv", "bill", "gus"],
      room_name: "room 1"
    },
    {
      users: ["edward", "marv", "bill", "gus"],
      room_name: "room 2"
    },
    {
      users: ["edward", "marv", "bill", "gus"],
      room_name: "room 3"
    }
  ];

  const test_username = "test_user";
  const test_roomname = "test_room";

  const expected_state_values = {
    username: test_username,
    gameRoom: test_roomname,
    data: test_player_details
  };

  it("Verify parameter provided PlayeValues are rendered correctly", () => {
    const component = shallow(
      <LobbyListDisplay
        currentUser={test_username}
        currentRoom={test_roomname}
        data={test_player_details}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("Verify state transfers via props as expected", () => {
    const wrapper = shallow(
      <LobbyListDisplay
        currentUser={test_username}
        currentRoom={test_roomname}
        data={test_player_details}
      />
    );
    expect(wrapper.state()).toEqual({ ...expected_state_values }); // passed
  });

  it("playerUpdateHandler mehthod updates state as expected", () => {
    const wrapper = shallow(
      <LobbyListDisplay
        currentUser={test_username}
        currentRoom={test_roomname}
        data={[]}
      />
    );

    const expected_state_values_with_no_player_details = {
      username: test_username,
      gameRoom: test_roomname,
      data: []
    };

    expect(wrapper.state()).toEqual({
      ...expected_state_values_with_no_player_details
    });

    const expected_backend_callback_message = {
      user: test_username,
      room: test_roomname,
      content: JSON.stringify(test_player_details)
    };

    wrapper.instance().playerUpdateHandler(expected_backend_callback_message);

    expect(wrapper.state()).toEqual({ ...expected_state_values }); // passed
  });
});
