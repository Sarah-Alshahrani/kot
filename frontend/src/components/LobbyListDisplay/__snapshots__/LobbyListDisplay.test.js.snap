// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify LobbyListDisplay layout template Verify parameter provided PlayeValues are rendered correctly 1`] = `
<div>
  <Bootstrap(Table)
    bordered={true}
    hover={true}
    striped={true}
  >
    <thead>
      <tr>
        <th>
          Players
        </th>
        <th>
          Room Name
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr
        key="0"
      >
        <td>
          edward
          marv
          bill
          gus
        </td>
        <td>
          room 1
        </td>
        <td>
          <a
            className="b"
            href="/game"
          >
            <button
              className="btn btn-secondary"
              id="room 1"
              onClick={[Function]}
              type="button"
            >
              Join Room 
              room 1
            </button>
          </a>
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
          edward
          marv
          bill
          gus
        </td>
        <td>
          room 2
        </td>
        <td>
          <a
            className="b"
            href="/game"
          >
            <button
              className="btn btn-secondary"
              id="room 2"
              onClick={[Function]}
              type="button"
            >
              Join Room 
              room 2
            </button>
          </a>
        </td>
      </tr>
      <tr
        key="2"
      >
        <td>
          edward
          marv
          bill
          gus
        </td>
        <td>
          room 3
        </td>
        <td>
          <a
            className="b"
            href="/game"
          >
            <button
              className="btn btn-secondary"
              id="room 3"
              onClick={[Function]}
              type="button"
            >
              Join Room 
              room 3
            </button>
          </a>
        </td>
      </tr>
    </tbody>
  </Bootstrap(Table)>
</div>
`;
